From be3093cd74d444f2756a6d6f5c4d1ebb04ff7f0a Mon Sep 17 00:00:00 2001
From: Jiebing Li <jiebing.li@intel.com>
Date: Fri, 1 Aug 2014 15:49:08 +0800
Subject: [PATCH] usb: dwc3-device: add OTG configuration for CHT platform

Intel Cherrytrail platform has it's own OTG PHY driver, so this patch
enables hooking of CHT OTG PHY driver in dwc3 driver, and add PCI id
for dwc3 controller on CHT platform.

Change-Id: Ide662ece5d876eea91c04a53d69af9bed806c2ba
Signed-off-by: geoffroy.weisenhorn <weisenhorn.geoffroy.@intel.com>
Signed-off-by: Jiebing Li <jiebing.li@intel.com>
Signed-off-by: Wu, Hao <hao.wu@intel.com>
(cherry picked from commit ed8888ed30392530707d780a59a19a552b62c155)

Conflicts manually resolved in:

	drivers/usb/dwc3/dwc3-pci.c
	drivers/usb/dwc3/gadget.c
---
 drivers/usb/dwc3/dwc3-pci.c |   13 +++++++++----
 drivers/usb/dwc3/gadget.c   |   10 ++++++++++
 2 files changed, 19 insertions(+), 4 deletions(-)

diff --git a/drivers/usb/dwc3/dwc3-pci.c b/drivers/usb/dwc3/dwc3-pci.c
index fc154d4..a2bfd05 100644
--- a/drivers/usb/dwc3/dwc3-pci.c
+++ b/drivers/usb/dwc3/dwc3-pci.c
@@ -32,8 +32,8 @@
 #define PCI_DEVICE_ID_SYNOPSYS_HAPSUSB3	0xabcd
 #define PCI_DEVICE_ID_INTEL_BYT		0x0f37
 #define PCI_DEVICE_ID_INTEL_MRFLD	0x119e
-#define PCI_DEVICE_ID_INTEL_BSW		0x22B7
-#define PCI_DEVICE_ID_INTEL_CGM		0x2BB7
+#define PCI_DEVICE_ID_INTEL_CHT		0x22b7
+#define PCI_DEVICE_ID_INTEL_CGM		0x2bb7
 
 struct dwc3_pci {
 	struct device		*dev;
@@ -54,7 +54,12 @@ static int dwc3_pci_register_phys(struct dwc3_pci *glue)
 
 	pci_dev = to_pci_dev(glue->dev);
 
-    pdev = platform_device_alloc("usb_phy_gen_xceiv", 0);
+	if (pci_dev->vendor == PCI_VENDOR_ID_INTEL &&
+			(pci_dev->device == PCI_DEVICE_ID_INTEL_CHT ||
+             pci_dev->device == PCI_DEVICE_ID_INTEL_CGM))
+		pdev = platform_device_alloc("intel-cht-otg", 0);
+	else
+		pdev = platform_device_alloc("usb_phy_gen_xceiv", 0);
 
 	if (!pdev)
 		return -ENOMEM;
@@ -211,9 +216,9 @@ static const struct pci_device_id dwc3_pci_id_table[] = {
 		PCI_DEVICE(PCI_VENDOR_ID_SYNOPSYS,
 				PCI_DEVICE_ID_SYNOPSYS_HAPSUSB3),
 	},
-	{ PCI_DEVICE(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_BSW), },
 	{ PCI_DEVICE(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_BYT), },
 	{ PCI_DEVICE(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_MRFLD), },
+	{ PCI_DEVICE(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_CHT), },
 	{ PCI_DEVICE(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_CGM), },
 	{  }	/* Terminating Entry */
 };
diff --git a/drivers/usb/dwc3/gadget.c b/drivers/usb/dwc3/gadget.c
index 4cd89d9..58df053 100644
--- a/drivers/usb/dwc3/gadget.c
+++ b/drivers/usb/dwc3/gadget.c
@@ -1485,6 +1485,16 @@ static int dwc3_gadget_start(struct usb_gadget *g,
 	int			irq;
 	u32			reg;
 
+	pm_runtime_get_sync(dwc->dev);
+
+	if (dwc->usb2_phy) {
+		ret = otg_set_peripheral(dwc->usb2_phy->otg, &dwc->gadget);
+		if (ret == -ENOTSUPP)
+			dev_info(dwc->dev, "no OTG driver registered\n");
+		else if (ret)
+			goto err0;
+	}
+
 	irq = platform_get_irq(to_platform_device(dwc->dev), 0);
 	ret = request_threaded_irq(irq, dwc3_interrupt, dwc3_thread_interrupt,
 			IRQF_SHARED, "dwc3", dwc);
-- 
1.7.9.5

